// @wdio/allure-reporter/build/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const reporter_1 = __importDefault(require("@wdio/reporter"));
const allure_js_commons_1 = __importDefault(require("allure-js-commons"));
const step_1 = __importDefault(require("allure-js-commons/beans/step"));
const utils_1 = require("./utils");
const constants_1 = require("./constants");
class AllureReporter extends reporter_1.default {
    constructor(options) {
        const outputDir = options.outputDir || 'allure-results';
        const useCucumberStepReporter = Boolean(options.useCucumberStepReporter);
        const disableMochaHooks = Boolean(options.disableMochaHooks);
        super({
            ...options,
            outputDir,
            useCucumberStepReporter,
            disableMochaHooks
        });
        this.config = {};
        this.capabilities = {};
        this.allure = new allure_js_commons_1.default();
        this.allure.setOptions({ targetDir: outputDir });
        this.registerListeners();
        this.lastScreenshot = undefined;
    }
    registerListeners() {
        process.on(constants_1.events.addLabel, this.addLabel.bind(this));
        process.on(constants_1.events.addFeature, this.addFeature.bind(this));
        process.on(constants_1.events.addStory, this.addStory.bind(this));
        process.on(constants_1.events.addSeverity, this.addSeverity.bind(this));
        process.on(constants_1.events.addIssue, this.addIssue.bind(this));
        process.on(constants_1.events.addTestId, this.addTestId.bind(this));
        process.on(constants_1.events.addEnvironment, this.addEnvironment.bind(this));
        process.on(constants_1.events.addAttachment, this.addAttachment.bind(this));
        process.on(constants_1.events.addDescription, this.addDescription.bind(this));
        process.on(constants_1.events.startStep, this.startStep.bind(this));
        process.on(constants_1.events.endStep, this.endStep.bind(this));
        process.on(constants_1.events.addStep, this.addStep.bind(this));
        process.on(constants_1.events.addArgument, this.addArgument.bind(this));
    }
    onRunnerStart(runner) {
        this.config = runner.config;
        this.capabilities = runner.capabilities;
        this.isMultiremote = runner.isMultiremote || false;
    }
    onSuiteStart(suite) {
        if (this.options.useCucumberStepReporter) {
            if (suite.type === 'feature') {
                return this.allure.startSuite(suite.title);
            }
            this.allure.startCase(suite.title);
            const currentTest = this.allure.getCurrentTest();
            this.getLabels(suite).forEach(({ name, value }) => {
                currentTest.addLabel(name, value);
            });
            if (suite.description) {
                this.addDescription(suite);
            }
            return this.setCaseParameters(suite.cid);
        }
        const currentSuite = this.allure.getCurrentSuite();
        const prefix = currentSuite ? currentSuite.name + ': ' : '';
        this.allure.startSuite(prefix + suite.title);
    }
    onSuiteEnd(suite) {
        if (this.options.useCucumberStepReporter && suite.type === 'scenario') {
            suite.hooks = suite.hooks.map(hook => {
                hook.state = hook.state ? hook.state : constants_1.stepStatuses.PASSED;
                return hook;
            });
            const suiteChildren = [...suite.tests, ...suite.hooks];
            const isPassed = !suiteChildren.some(item => item.state !== constants_1.testStatuses.PASSED);
            if (isPassed) {
                return this.allure.endCase(constants_1.testStatuses.PASSED);
            }
            const isSkipped = suiteChildren.every(item => [constants_1.stepStatuses.PASSED, constants_1.stepStatuses.SKIPPED].indexOf(item.state) >= 0);
            if (isSkipped) {
                return this.allure.endCase(constants_1.testStatuses.PENDING);
            }
            return;
        }
        this.allure.endSuite();
    }
    onTestStart(test) {
        const testTitle = test.currentTest ? test.currentTest : test.title;
        if (this.isAnyTestRunning() && this.allure.getCurrentTest().name == testTitle) {
            this.setCaseParameters(test.cid);
            return;
        }
        if (this.options.useCucumberStepReporter) {
            return this.allure.startStep(testTitle);
        }
        this.allure.startCase(testTitle);
        this.setCaseParameters(test.cid);
    }
    setCaseParameters(cid) {
        const currentTest = this.allure.getCurrentTest();
        if (!this.isMultiremote) {
            const { browserName, deviceName, desired, device } = this.capabilities;
            let targetName = device || browserName || deviceName || cid;
            if (desired && desired.deviceName && desired.platformVersion) {
                targetName = `${device || desired.deviceName} ${desired.platformVersion}`;
            }
            const browserstackVersion = this.capabilities.os_version || this.capabilities.osVersion;
            const version = browserstackVersion || this.capabilities.browserVersion || this.capabilities.version || this.capabilities.platformVersion || '';
            const paramName = (deviceName || device) ? 'device' : 'browser';
            const paramValue = version ? `${targetName}-${version}` : targetName;
            currentTest.addParameter('argument', paramName, paramValue);
        }
        else {
            currentTest.addParameter('argument', 'isMultiremote', 'true');
        }
        currentTest.addLabel('language', 'javascript');
        currentTest.addLabel('framework', 'wdio');
        currentTest.addLabel('thread', cid);
    }
    getLabels({ tags }) {
        const labels = [];
        if (tags) {
            tags.forEach((tag) => {
                const label = tag.name.replace(/[@]/, '').split('=');
                if (label.length === 2) {
                    labels.push({ name: label[0], value: label[1] });
                }
            });
        }
        return labels;
    }
    onTestPass() {
        if (this.options.useCucumberStepReporter) {
            return this.allure.endStep(constants_1.stepStatuses.PASSED);
        }
        this.allure.endCase(constants_1.testStatuses.PASSED);
    }
    onTestFail(test) {
        if (this.options.useCucumberStepReporter) {
            const testStatus = utils_1.getTestStatus(test, this.config);
            const stepStatus = Object.values(constants_1.stepStatuses).indexOf(testStatus >= 0) ?
                testStatus : constants_1.stepStatuses.FAILED;
            this.allure.endStep(stepStatus);
            this.allure.endCase(testStatus, utils_1.getErrorFromFailedTest(test));
            return;
        }
        if (!this.isAnyTestRunning()) {
            this.onTestStart(test);
        }
        else {
            this.allure.getCurrentTest().name = test.title;
        }
        const status = utils_1.getTestStatus(test, this.config);
        while (this.allure.getCurrentSuite().currentStep instanceof step_1.default) {
            this.allure.endStep(status);
        }
        this.allure.endCase(status, utils_1.getErrorFromFailedTest(test));
    }
    onTestSkip(test) {
        if (this.options.useCucumberStepReporter) {
            this.allure.endStep(constants_1.stepStatuses.CANCELED);
        }
        else if (!this.allure.getCurrentTest() || this.allure.getCurrentTest().name !== test.title) {
            this.allure.pendingCase(test.title);
        }
        else {
            this.allure.endCase(constants_1.testStatuses.PENDING);
        }
    }
    onBeforeCommand(command) {
        if (!this.isAnyTestRunning()) {
            return;
        }
        const { disableWebdriverStepsReporting } = this.options;
        if (disableWebdriverStepsReporting || this.isMultiremote) {
            return;
        }
        this.allure.startStep(command.method
            ? `${command.method} ${command.endpoint}`
            : command.command);
        const payload = command.body || command.params;
        if (!utils_1.isEmpty(payload)) {
            this.dumpJSON('Request', payload);
        }
    }
    onAfterCommand(command) {
        const { disableWebdriverStepsReporting, disableWebdriverScreenshotsReporting } = this.options;
        if (this.isScreenshotCommand(command) && command.result.value) {
            if (!disableWebdriverScreenshotsReporting) {
                this.lastScreenshot = command.result.value;
            }
        }
        if (!this.isAnyTestRunning()) {
            return;
        }
        this.attachScreenshot();
        if (this.isMultiremote) {
            return;
        }
        if (!disableWebdriverStepsReporting) {
            if (command.result && command.result.value && !this.isScreenshotCommand(command)) {
                this.dumpJSON('Response', command.result.value);
            }
            const suite = this.allure.getCurrentSuite();
            if (!suite || !(suite.currentStep instanceof step_1.default)) {
                return;
            }
            this.allure.endStep(constants_1.testStatuses.PASSED);
        }
    }
    onHookStart(hook) {
        if (!hook.parent || !this.allure.getCurrentSuite()) {
            return false;
        }
        if (this.options.disableMochaHooks && utils_1.isMochaEachHooks(hook.title)) {
            if (this.allure.getCurrentTest()) {
                this.allure.startStep(hook.title);
            }
            return;
        }
        if (this.options.disableMochaHooks && utils_1.isMochaAllHooks(hook.title)) {
            return;
        }
        this.onTestStart(hook);
    }
    onHookEnd(hook) {
        if (!hook.parent || !this.allure.getCurrentSuite() || (this.options.disableMochaHooks && !utils_1.isMochaAllHooks(hook.title) && !this.allure.getCurrentTest())) {
            return false;
        }
        if (this.options.disableMochaHooks && utils_1.isMochaEachHooks(hook.title)) {
            if (hook.error) {
                this.allure.endStep(constants_1.stepStatuses.FAILED);
            }
            else {
                this.allure.endStep(constants_1.stepStatuses.PASSED);
            }
            return;
        }
        if (hook.error) {
            if (this.options.disableMochaHooks && utils_1.isMochaAllHooks(hook.title)) {
                this.onTestStart(hook);
                this.attachScreenshot();
            }
            this.onTestFail(hook);
        }
        else if (this.options.disableMochaHooks || this.options.useCucumberStepReporter) {
            if (!utils_1.isMochaAllHooks(hook.title)) {
                this.onTestPass();
                if (this.allure.getCurrentTest().steps.length === 0 && !this.options.useCucumberStepReporter) {
                    this.allure.getCurrentSuite().testcases.pop();
                }
                else if (this.options.useCucumberStepReporter) {
                    const step = this.allure.getCurrentTest().steps.pop();
                    if (step && step.attachments.length >= 1) {
                        step.attachments.forEach(attachment => {
                            this.allure.getCurrentTest().addAttachment(attachment);
                        });
                    }
                }
            }
        }
        else if (!this.options.disableMochaHooks)
            this.onTestPass();
    }
    addLabel({ name, value }) {
        if (!this.isAnyTestRunning()) {
            return false;
        }
        const test = this.allure.getCurrentTest();
        test.addLabel(name, value);
    }
    addStory({ storyName }) {
        if (!this.isAnyTestRunning()) {
            return false;
        }
        const test = this.allure.getCurrentTest();
        test.addLabel('story', storyName);
    }
    addFeature({ featureName }) {
        if (!this.isAnyTestRunning()) {
            return false;
        }
        const test = this.allure.getCurrentTest();
        test.addLabel('feature', featureName);
    }
    addSeverity({ severity }) {
        if (!this.isAnyTestRunning()) {
            return false;
        }
        const test = this.allure.getCurrentTest();
        test.addLabel('severity', severity);
    }
    addIssue({ issue }) {
        if (!this.isAnyTestRunning()) {
            return false;
        }
        const test = this.allure.getCurrentTest();
        const issueLink = utils_1.getLinkByTemplate(this.options.issueLinkTemplate, issue);
        test.addLabel('issue', issueLink);
    }
    addTestId({ testId }) {
        if (!this.isAnyTestRunning()) {
            return false;
        }
        const test = this.allure.getCurrentTest();
        const tmsLink = utils_1.getLinkByTemplate(this.options.tmsLinkTemplate, testId);
        test.addLabel('testId', tmsLink);
    }
    addEnvironment({ name, value }) {
        if (!this.isAnyTestRunning()) {
            return false;
        }
        const test = this.allure.getCurrentTest();
        test.addParameter('environment-variable', name, value);
    }
    addDescription({ description, descriptionType }) {
        if (!this.isAnyTestRunning()) {
            return false;
        }
        const test = this.allure.getCurrentTest();
        test.setDescription(description, descriptionType);
    }
    addAttachment({ name, content, type = 'text/plain' }) {
        if (!this.isAnyTestRunning()) {
            return false;
        }
        if (type === 'application/json') {
            this.dumpJSON(name, content);
        }
        else {
            this.allure.addAttachment(name, Buffer.from(content), type);
        }
    }
    startStep(title) {
        if (!this.isAnyTestRunning()) {
            return false;
        }
        this.allure.startStep(title);
    }
    endStep(status) {
        if (!this.isAnyTestRunning()) {
            return false;
        }
        this.allure.endStep(status);
    }
    addStep({ step }) {
        if (!this.isAnyTestRunning()) {
            return false;
        }
        this.startStep(step.title);
        if (step.attachment) {
            this.addAttachment(step.attachment);
        }
        this.endStep(step.status);
    }
    addArgument({ name, value }) {
        if (!this.isAnyTestRunning()) {
            return false;
        }
        const test = this.allure.getCurrentTest();
        test.addParameter('argument', name, value);
    }
    isAnyTestRunning() {
        return this.allure.getCurrentSuite() && this.allure.getCurrentTest();
    }
    isScreenshotCommand(command) {
        const isScrenshotEndpoint = /\/session\/[^/]*(\/element\/[^/]*)?\/screenshot/;
        return (isScrenshotEndpoint.test(command.endpoint) ||
            command.command === 'takeScreenshot');
    }
    dumpJSON(name, json) {
        const content = JSON.stringify(json, null, 2);
        const isStr = typeof content === 'string';
        this.allure.addAttachment(name, isStr ? content : `${content}`, isStr ? 'application/json' : 'text/plain');
    }
    attachScreenshot() {
        if (this.lastScreenshot && !this.options.disableWebdriverScreenshotsReporting) {
            this.allure.addAttachment('Screenshot', Buffer.from(this.lastScreenshot, 'base64'));
            this.lastScreenshot = undefined;
        }
    }
}
AllureReporter.addFeature = (featureName) => {
    utils_1.tellReporter(constants_1.events.addFeature, { featureName });
};
AllureReporter.addLabel = (name, value) => {
    utils_1.tellReporter(constants_1.events.addLabel, { name, value });
};
AllureReporter.addSeverity = (severity) => {
    utils_1.tellReporter(constants_1.events.addSeverity, { severity });
};
AllureReporter.addIssue = (issue) => {
    utils_1.tellReporter(constants_1.events.addIssue, { issue });
};
AllureReporter.addTestId = (testId) => {
    utils_1.tellReporter(constants_1.events.addTestId, { testId });
};
AllureReporter.addStory = (storyName) => {
    utils_1.tellReporter(constants_1.events.addStory, { storyName });
};
AllureReporter.addEnvironment = (name, value) => {
    utils_1.tellReporter(constants_1.events.addEnvironment, { name, value });
};
AllureReporter.addDescription = (description, descriptionType) => {
    utils_1.tellReporter(constants_1.events.addDescription, { description, descriptionType });
};
AllureReporter.addAttachment = (name, content, type) => {
    if (!type) {
        type = content instanceof Buffer ? 'image/png' : typeof content === 'string' ? 'text/plain' : 'application/json';
    }
    utils_1.tellReporter(constants_1.events.addAttachment, { name, content, type });
};
AllureReporter.startStep = (title) => {
    utils_1.tellReporter(constants_1.events.startStep, title);
};
AllureReporter.endStep = (status = constants_1.stepStatuses.PASSED) => {
    if (!Object.values(constants_1.stepStatuses).includes(status)) {
        throw new Error(`Step status must be ${Object.values(constants_1.stepStatuses).join(' or ')}. You tried to set "${status}"`);
    }
    utils_1.tellReporter(constants_1.events.endStep, status);
};
AllureReporter.addStep = (title, { content, name = 'attachment', type = 'text/plain' } = {}, status = constants_1.stepStatuses.PASSED) => {
    if (!Object.values(constants_1.stepStatuses).includes(status)) {
        throw new Error(`Step status must be ${Object.values(constants_1.stepStatuses).join(' or ')}. You tried to set "${status}"`);
    }
    const step = content ? { title, attachment: { content, name, type }, status } : { title, status };
    utils_1.tellReporter(constants_1.events.addStep, { step });
};
AllureReporter.addArgument = (name, value) => {
    utils_1.tellReporter(constants_1.events.addArgument, { name, value });
};
exports.default = AllureReporter;
